### ISSUE TITLE
Generated UFOs missing weight in fontinfo (fontmake 3.4.0)

### ISSUE THREAD

**Issue by rsms** [googlefonts/fontmake#943]
`fontmake -o ufo -g Inter-Roman.glyphs` produces UFOs which are missing `openTypeOS2WeightClass` entries in fontinfo. The generated designspace file is correct and contains weight data.

Repro: (unzip attached glyphs source: [Inter-Roman.glyphs.zip](https://github.com/googlefonts/fontmake/files/9717422/Inter-Roman.glyphs.zip))

```
$ fontmake -o ufo -g Inter-Roman.glyphs
INFO:fontmake.font_project:Building master UFOs and designspace from Glyphs source
INFO:glyphsLib.classes:Parsing "Inter-Roman.glyphs" file into <GSFont>
$ grep -i weight master_ufo/Inter-*.ufo/fontinfo.plist
# no results
$ grep -i weight master_ufo/Inter.designspace
    <axis tag="wght" name="Weight" minimum="100" maximum="900" default="400"/>
        <dimension name="Weight" xvalue="100"/>
        <dimension name="Weight" xvalue="100"/>
        <dimension name="Weight" xvalue="400"/>
        <dimension name="Weight" xvalue="400"/>
        <dimension name="Weight" xvalue="900"/>
        <dimension name="Weight" xvalue="900"/>
        <dimension name="Weight" xvalue="100"/>
        <dimension name="Weight" xvalue="200"/>
        <dimension name="Weight" xvalue="300"/>
        <dimension name="Weight" xvalue="400"/>
        <dimension name="Weight" xvalue="500"/>
        <dimension name="Weight" xvalue="600"/>
        <dimension name="Weight" xvalue="700"/>
        <dimension name="Weight" xvalue="800"/>
        <dimension name="Weight" xvalue="900"/>
        <dimension name="Weight" xvalue="100"/>
        <dimension name="Weight" xvalue="200"/>
        <dimension name="Weight" xvalue="300"/>
        <dimension name="Weight" xvalue="400"/>
        <dimension name="Weight" xvalue="500"/>
        <dimension name="Weight" xvalue="600"/>
        <dimension name="Weight" xvalue="700"/>
        <dimension name="Weight" xvalue="800"/>
        <dimension name="Weight" xvalue="900"/>
$
```

```
$ fontmake --version
3.4.0
$ python --version
Python 3.10.6
```

**Comment by anthrotype** [comment]
cc @jenskutilek who is currently touching that code in https://github.com/googlefonts/glyphsLib/pull/810, he may have some insights

**Comment by jenskutilek** [comment]
> I checked the .glyphs file (in the Glyphs app) and indeed all masters have "Axis Location" (called "Axes Coordinates" in the GUI:)

No, that’s not the Axis Location. Axis Location is a custom parameter on the masters and/or instances.

In your case, the user locations are derived from the master and instance coordinates.

It seems that the correct weight and width class values are only written to the master UFOs if the font indeed has "Axis Location" parameters on all masters: https://github.com/googlefonts/glyphsLib/blob/main/Lib/glyphsLib/builder/masters.py#L66

I think that's a bug in glyphsLib…

**Comment by anthrotype** [comment]
> I think that's a bug in glyphsLib…

I'm not sure it's a bug though, if the VF or the static instances have that OS/2 weight class set correctly, that's all you need. The master UFOs are not meant to be exported to TTF/OTF as such, they are intermediate build products. Even from Glyphs.app one can only export a VF or a static instance, never a "master" TTF/OTF. I know fontmake still allows one to build TTF/OTF from the masters directly, but I'd argue that's a legacy behavior that we may wish to get rid of eventually.

**Comment by jenskutilek** [comment]
Btw. why doesn't `fontmake -o otf -g Inter-Roman.glyphs` generate instance OTFs? Is that expected?

**Comment by anthrotype** [comment]
yes, it's the normal fontmake's behavior since day one, it will build master TTF/OTF files. I think it's a legacy from noto projects that were not interpolatable and as such only master needed to be exported.

**Comment by anthrotype** [comment]
To build the instances, as you know, you need to pass the `-i` option. If an instance is the same as a master, I think (I hope!) that it won't try to do the interpolation (with the risk of failing to interpolate if masters aren't compatible), but will only apply the instance-specific custom parameters.

**Comment by anthrotype** [comment]
> I think (I hope!) that it won't try to do the interpolation (with the risk of failing to interpolate if masters aren't compatible)

would be nice to confirm that, I honestly don't remember

**Comment by jenskutilek** [comment]
(I deleted my comment about fontmake only building master OTFs/UFOs. I found the -i option. To my defense, I've only been using fontmake to build variable fonts ;))

**Comment by rsms** [comment]
> So is there an issue after all? Or is the way @rsms builds his static fonts just unsupported?

I'm building 72 static fonts (36 otf + 36 ttf) and I use make to manage the build ([there are many steps involved](https://github.com/rsms/inter/blob/2ca005973c364b584019bade7ff2c3eb6c4521c3/Makefile) beyond just building static fonts.)

There are a couple of reasons building OTFs directly from glyphs files is not a good option for me:
- It is MUCH slower (see below)
- I need to do processing/patching of the UFOs before building fonts. [postprocess-designspace.py](https://github.com/rsms/inter/blob/2ca005973c364b584019bade7ff2c3eb6c4521c3/misc/tools/postprocess-designspace.py) does the following:
  - Correction of bad opsz and wght range
  - Updates version (from current git tree hash + version.txt file)
  - Marks composed glyphs with complex transformations for decomposition

So ideally I'd like to stick with building OTFs & TTFs from UFOs rather than .glyphs.

When I build with `fontmake -g ...glyphs -o otf -i ...` the build is much slower (twice as slow compared to building from UFO.) Compiling with fontmake is already very slow: ~15 second to compile one static OTF from a UFO — that's longer than it takes to compile a medium-sized C project on my machine. Currently [it takes about 11 minutes](https://github.com/rsms/inter/actions/runs/3207844717) to build the font on GitHub CI.

Building from UFO: (16s)

```
$ time fontmake -u Inter-Black.ufo -o otf --output-path InterBlack2.otf
INFO:fontmake.font_project:Building OTF for Inter-Black
INFO:ufo2ft:Pre-processing glyphs
INFO:ufo2ft.filters.base:Running EraseOpenCornersFilter on Inter-Black
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on Inter-Black
INFO:ufo2ft.filters.base:Running RemoveOverlapsFilter on Inter-Black
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on Inter-Black
INFO:ufo2ft:Building OpenType tables
INFO:ufo2ft.postProcessor:Subroutinizing CFF table with cffsubr
INFO:ufo2ft.postProcessor:Renaming glyphs to final production names
INFO:fontmake.font_project:Saving InterBlack2.otf
16.07s user 0.20s system 99% cpu 16.278 total
```

Building from glyphs file: (36s)

```
$ time fontmake -g Inter-Roman.glyphs -o otf -i "Inter Black" --output-path InterBlack.otf
INFO:fontmake.font_project:Building master UFOs and designspace from Glyphs source
INFO:glyphsLib.classes:Parsing "Inter-Roman.glyphs" file into <GSFont>
INFO:fontmake.font_project:Interpolating master UFOs from designspace
INFO:fontmake.font_project:Generating instance UFO for "Inter Black"
INFO:fontmake.font_project:Building OTF for Inter-Black
INFO:ufo2ft:Pre-processing glyphs
INFO:ufo2ft.filters.base:Running EraseOpenCornersFilter on Inter-Black
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on Inter-Black
INFO:ufo2ft.filters.base:Running RemoveOverlapsFilter on Inter-Black
INFO:ufo2ft:Building OpenType tables
INFO:ufo2ft.postProcessor:Subroutinizing CFF table with cffsubr
INFO:ufo2ft.postProcessor:Renaming glyphs to final production names
INFO:fontmake.font_project:Saving InterBlack.otf
35.87s user 2.93s system 99% cpu 39.046 total
```

I'm currently working around the missing weight info in the UFOs by updating the UFOs after they are generated:
 https://github.com/rsms/inter/blob/2ca005973c364b584019bade7ff2c3eb6c4521c3/misc/tools/postprocess-designspace.py#L80


> No, that’s not the Axis Location. Axis Location is a custom parameter on the masters and/or instances.

Interesting! Today I learned a new thing :-)

**Comment by anthrotype** [comment]
It's totally fine that you split the build phases and build from UFOs, what we mean is when generating the UFOs for the static builds you need to pass the -i option so that fontmake will generate instances, rather than masters, and those will contain all the necessary metadata.

**Comment by anthrotype** [comment]
`fontmake Inter.glyphs -o ufo -i` will generate instance UFOs, by default in the `instance_ufo` dir, but can change that with --instance-dir

**Comment by rsms** [comment]
> It's totally fine that you split the build phases and build from UFOs, what we mean is when generating the UFOs for the static builds you need to pass the -i option so that fontmake will generate instances, rather than masters, and those will contain all the necessary metadata.

Currently fontmake generates UFOs for each master unconditionally when generating the designspace (which I need for VF and which needs patching.) Producing the UFOs twice will obviously be slower but only about 15s (about 10% speed penalty.) Ideally fontmake would have an option to write a designspace file without writing master UFOs. That would speed things up!
